samples/apps/logging/logging.cpp

```diff
@@ -581,6 +582,140 @@ namespace loggingapp
+          return;
+        }
+
+        // Use hash of request as RequestHandle. WARNING: This means identical
+        // requests from different users will collide, and overwrite each
+        // other's progress!
+        ccf::historical::RequestHandle handle;
+        {
+          auto size =
+            sizeof(in.from_seqno) + sizeof(in.to_seqno) + sizeof(in.id);
+          std::vector<uint8_t> v(size);
+          auto data = v.data();
+          serialized::write(data, size, in.from_seqno);
+          serialized::write(data, size, in.to_seqno);
+          serialized::write(data, size, in.id);
+          handle = std::hash<decltype(v)>()(v);
+        }
+
+        // Fetch the requested range
+        auto& historical_cache = context.get_historical_state();
+
+        // TODO: Add pagination limit
+
+        auto stores =
+          historical_cache.get_store_range(handle, in.from_seqno, in.to_seqno);
+        if (stores.empty())
+        {
+          args.rpc_ctx->set_response_status(HTTP_STATUS_ACCEPTED);
+          static constexpr size_t retry_after_seconds = 3;
+          args.rpc_ctx->set_response_header(
+            http::headers::RETRY_AFTER, retry_after_seconds);
+          args.rpc_ctx->set_response_header(
+            http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+          args.rpc_ctx->set_response_body(fmt::format(
+            "Historical transactions from {} to {} are not yet "
+            "available, fetching now",
+            in.from_seqno,
+            in.to_seqno));
+          return;
+        }
+
+        // Process the fetched Stores
+        LoggingGetHistoricalRange::Out response;
+        for (size_t i = 0; i < stores.size(); ++i)
+        {
+          const auto store_seqno = in.from_seqno + i;
+          auto& store = stores[i];
+
+          auto historical_tx = store->create_read_only_tx();
+          auto records_handle = historical_tx.ro(records);
+          const auto v = records_handle->get(in.id);
+
+          if (v.has_value())
+          {
+            LoggingGetHistoricalRange::Entry e;
+            e.seqno = store_seqno;
+            e.id = in.id;
+            e.msg = v.value();
+            response.entries.push_back(e);
+          }
+          // We do not include an entry when the given key wasn't modified at
+          // this seqno!
+        }
+
+        // Construct the HTTP response
+        nlohmann::json j_response = response;
+        args.rpc_ctx->set_response_status(HTTP_STATUS_OK);
+        args.rpc_ctx->set_response_header(
+          http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+        args.rpc_ctx->set_response_body(j_response.dump());
+
+        // ALSO: Assume this response makes it all the way to the client, and
+        // they're finished with it, so we can drop the retrieved state. In a
+        // real app this should be driven by a separate client request!
+        historical_cache.drop_request(handle);
```

Comment:
I think it's more likely that a real app would want a cap of some sort on the historical cache size, and let stuff drop when either that or the expiry time hits.

Reply:
I agree. I've extended the comment to suggest this. I was aiming to show here "you should think about and call this somewhere", but also "this approach is a bodge for this demo and not something you should copy verbatim".